ios/Gekidou/Sources/Gekidou/PushNotification/PushNotification+Signature.swift

```diff
@@ -0,0 +1,133 @@
+        return self.verifySignature(notification.userInfo)
+    }
+    public func verifySignature(_ userInfo: [AnyHashable : Any]) -> Bool {
+        guard let signature =  userInfo["signature"] as? String
+        else {
+            os_log(
+                OSLogType.default,
+                "Mattermost Notifications: Signature verification: No signature in the notification"
+            )
+            return true
+        }
+        
+        guard let serverId =  userInfo["server_id"] as? String
+        else {
+            os_log(
+                OSLogType.default,
+                "Mattermost Notifications: Signature verification: No server_id in the notification"
+            )
+            return false
+        }
+        
+        guard let serverUrl = try? Database.default.getServerUrlForServer(serverId)
+        else {
+            os_log(
+                OSLogType.default,
+                "Mattermost Notifications: Signature verification: No server_url for server_id"
+            )
+            return false
+        }
+        
+        if signature == "NO_SIGNATURE" {
+            guard let version = Database.default.getConfig(serverUrl, "Version")
+            else {
+                os_log(
+                    OSLogType.default,
+                    "Mattermost Notifications: Signature verification: No server version"
+                )
+                return false
+            }
+            
+            // Verify server version
+        }
+        
+        guard let signingKey = Database.default.getConfig(serverUrl, "AsymmetricSigningPublicKey")
+        else {
+            os_log(
+                OSLogType.default,
+                "Mattermost Notifications: Signature verification: No signing key"
+            )
+            return false
+        }
+
+        let keyPEM = """
+-----BEGIN PUBLIC KEY-----
+\(signingKey)
+-----END PUBLIC KEY-----
+"""
+        let jwtVerifier = JWTVerifier.es256(publicKey: keyPEM.data(using: .utf8)!)
+        guard let newJWT = try? JWT<NotificationClaims>(jwtString: signature, verifier: jwtVerifier)
+        else {
+            os_log(
+                OSLogType.default,
+                "Mattermost Notifications: Signature verification: Cannot verify the signature"
+            )
+            return false
+        }
+        
+        guard let ackId = userInfo["ack_id"] as? String
+        else {
+            os_log(
+                OSLogType.default,
+                "Mattermost Notifications: Signature verification: No ack_id in the notification"
+            )
+            return false
+        }
+        
+        if (ackId != newJWT.claims.ack_id) {
+            os_log(
+                OSLogType.default,
+                "Mattermost Notifications: Signature verification: ackId is different"
+            )
+            return false
+        }
+        
+        guard let storedDeviceToken = Database.default.getDeviceToken()
```

Comment:
The part about the device Id seems totally irrelevant to me, no notification should be received if the wrong deviceId was used...

Can you please clarify why this is being checked?

Reply:
Here is the reasoning to use also the device id: https://docs.google.com/document/d/19TxvZErU-vbX8eOm0Uk8SI8WlheJlVTYcByzGzROK-c/edit?disco=AAABFuWaBis